sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/ReduceNumShufflePartitions.scala

```diff
@@ -82,7 +82,12 @@ case class ReduceNumShufflePartitions(conf: SQLConf) extends Rule[SparkPlan] {
       // `ShuffleQueryStageExec` gives null mapOutputStatistics when the input RDD has 0 partitions,
       // we should skip it when calculating the `partitionStartIndices`.
       val validMetrics = shuffleMetrics.filter(_ != null)
-      if (validMetrics.nonEmpty) {
+      // We may have different pre-shuffle partition numbers, don't reduce shuffle partition number
+      // in that case. For example when we union fully aggregated data (data is arranged to a single
+      // partition) and a result of a SortMergeJoin (multiple partitions).
+      val distinctNumPreShufflePartitions =
+        validMetrics.map(stats => stats.bytesByPartitionId.length).distinct
+      if (validMetrics.nonEmpty && distinctNumPreShufflePartitions.length == 1) {
```

Comment:
If we have this condition `distinctNumPreShufflePartitions.length == 1`, why do we need the assert at L136? Shall we remove the assert?

Reply:
Yes, we could remove it, but this assert has been there since the original version of `ReduceNumShufflePartitions` where the `distinctNumPreShufflePartitions.length == 1` check was also included. I'm not sure what is the plan with `ReduceNumShufflePartitions`. @carsonwang, @maryannxue do you want to improve `Union`/`SinglePartition` handling in it?

Reply:
I think it is fine to remove it. We can improve the handling of `Union/SinglePartition` in future and it probably needs more changes and a new function to estimate the partition start indices.